### ISSUE TITLE
v3.10 upgrade fails on task Migrate from etcd to CustomResources

### ISSUE THREAD

**Issue by felsys** [openshift/openshift-ansible#9750]
#### Description

Upgrading from 3.9.0 > 3.10.0 on Atomic host fails on the wait task following the "Migrate from etcd to CustomResources" task

##### Version

```
ansible==2.6.3
openshift-ansible-3.10.35-1-11-ga8d56b0
```

##### Steps To Reproduce
1. Run 3.10 upgrade playbook on Atomic host

##### Expected Results
Success

##### Observed Results

```
TASK [ansible_service_broker : Fail out because the ASB etcd to CRD migration was unsuccessful] *************************************************************************
fatal: [openshift-master]: FAILED! => {"changed": false, "msg": "The migration from etcd to CustomResourceDefinitions was not successful, aborting upgrade of the ansible service broker.\n"}
```

##### Additional Information

I inspected the openshift-ansible-service-broker project and I see 4 asb-etcd-migration-(hash) pods, all with the failure log message of:

```
container_linux.go:247: starting container process caused "exec: \"/usr/bin/migration\": stat /usr/bin/migration: no such file or directory"
```

**Comment by fabianvf** [comment]
What version of the broker container is deployed? My guess would be it's an old one that doesn't have the migration present... Not sure why that would only show up on atomic though.

**Comment by feld-ims** [comment]
I'm not entirely sure how to extract that version, but 'docker images' on the nodes show the versions of docker.io/ansibleplaybookbundle/origin-ansible-service-broker is over 5 months old (latest, v3.9, and v3.7)

**Comment by feld-ims** [comment]
Ah, that seems to be the issue, line 84 in ../roles/ansible_service_broker/tasks/migrate.yml

`imagePullPolicy: IfNotPresent`

Maybe that should be changed to always

**Comment by fabianvf** [comment]
I think we default to IfNotPresent for the sake of disconnected installs, but when upgrading from 3.9 to 3.10 the image tag should change from 3.9 to 3.10, so it should pull regardless... Unless the image tag is latest, in which case it would probably not update, that would explain this behavior. Can you paste the output of 

    oc get dc asb -n openshift-ansible-service-broker -o yaml

**Comment by feld-ims** [comment]
do you want the results of that dc output command pre or post (failed) upgrade?

**Comment by fabianvf** [comment]
Post is what I intended, but both would be helpful if it's not too much bother.

**Comment by feld-ims** [comment]
Looks like the v3.10 upgrade is still using the 3.9 image?

**PRE:**
```apiVersion: apps.openshift.io/v1
kind: DeploymentConfig
metadata:
  creationTimestamp: 2018-08-21T16:42:44Z
  generation: 2
  labels:
    app: openshift-ansible-service-broker
    service: asb
  name: asb
  namespace: openshift-ansible-service-broker
  resourceVersion: "15730"
  selfLink: /apis/apps.openshift.io/v1/namespaces/openshift-ansible-service-broker/deploymentconfigs/asb
  uid: 3a2c9eb7-a561-11e8-aa9e-005056b739a8
spec:
  replicas: 1
  selector:
    app: openshift-ansible-service-broker
  strategy:
    activeDeadlineSeconds: 21600
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: openshift-ansible-service-broker
        service: asb
    spec:
      containers:
      - env:
        - name: BROKER_CONFIG
          value: /etc/ansible-service-broker/config.yaml
        - name: HTTP_PROXY
        - name: HTTPS_PROXY
        - name: NO_PROXY
        image: ansibleplaybookbundle/origin-ansible-service-broker:v3.9
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 1338
            scheme: HTTPS
          initialDelaySeconds: 15
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: asb
        ports:
        - containerPort: 1338
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 1338
            scheme: HTTPS
          initialDelaySeconds: 15
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /tmp/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/ansible-service-broker
          name: config-volume
        - mountPath: /etc/tls/private
          name: asb-tls
        - mountPath: /var/run/asb-etcd-auth
          name: asb-etcd-auth
      dnsPolicy: ClusterFirst
      nodeSelector:
        region: infra
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: asb
      serviceAccountName: asb
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: broker-config
            path: config.yaml
          name: broker-config
        name: config-volume
      - name: asb-tls
        secret:
          defaultMode: 420
          secretName: asb-tls
      - name: asb-etcd-auth
        secret:
          defaultMode: 420
          secretName: broker-etcd-auth-secret
  test: false
  triggers:
  - type: ConfigChange
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2018-08-27T18:22:00Z
    lastUpdateTime: 2018-08-27T18:22:00Z
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  - lastTransitionTime: 2018-08-27T18:21:11Z
    lastUpdateTime: 2018-08-27T18:22:01Z
    message: replication controller "asb-3" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  details:
    causes:
    - type: Manual
    message: manual change
  latestVersion: 3
  observedGeneration: 2
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1
```

**POST:**
```apiVersion: apps.openshift.io/v1
kind: DeploymentConfig
metadata:
  creationTimestamp: 2018-08-21T16:42:44Z
  generation: 4
  labels:
    app: openshift-ansible-service-broker
    service: asb
  name: asb
  namespace: openshift-ansible-service-broker
  resourceVersion: "22508"
  selfLink: /apis/apps.openshift.io/v1/namespaces/openshift-ansible-service-broker/deploymentconfigs/asb
  uid: 3a2c9eb7-a561-11e8-aa9e-005056b739a8
spec:
  replicas: 1
  selector:
    app: openshift-ansible-service-broker
  strategy:
    activeDeadlineSeconds: 21600
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: openshift-ansible-service-broker
        service: asb
    spec:
      containers:
      - env:
        - name: BROKER_CONFIG
          value: /etc/ansible-service-broker/config.yaml
        - name: HTTP_PROXY
        - name: HTTPS_PROXY
        - name: NO_PROXY
        image: ansibleplaybookbundle/origin-ansible-service-broker:v3.9
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 1338
            scheme: HTTPS
          initialDelaySeconds: 15
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: asb
        ports:
        - containerPort: 1338
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 1338
            scheme: HTTPS
          initialDelaySeconds: 15
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /tmp/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/ansible-service-broker
          name: config-volume
        - mountPath: /etc/tls/private
          name: asb-tls
        - mountPath: /var/run/asb-etcd-auth
          name: asb-etcd-auth
      dnsPolicy: ClusterFirst
      nodeSelector:
        region: infra
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: asb
      serviceAccountName: asb
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: broker-config
            path: config.yaml
          name: broker-config
        name: config-volume
      - name: asb-tls
        secret:
          defaultMode: 420
          secretName: asb-tls
      - name: asb-etcd-auth
        secret:
          defaultMode: 420
          secretName: broker-etcd-auth-secret
  test: false
  triggers:
  - type: ConfigChange
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2018-08-27T18:21:11Z
    lastUpdateTime: 2018-08-27T18:22:01Z
    message: replication controller "asb-3" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: 2018-08-27T19:11:12Z
    lastUpdateTime: 2018-08-27T19:11:12Z
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  details:
    causes:
    - type: Manual
    message: manual change
  latestVersion: 3
  observedGeneration: 4
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1```

**Comment by jekader** [comment]
I hit this issue as well. Is there a known workaround to unblock the upgrade? Is it safe to just update the deployment config manually and specify the new image version?

**Comment by feld-ims** [comment]
The issue seems to be with it using the "latest" image tag and an older image still present on the nodes, and the ansible docker pull policy "IfNotPresent". While I can't upgrade and test due to another issue at the moment, you may be able to get around this by manually deleting the image on all of the nodes before the upgrade, or running "docker system prune --all" on all of the nodes first. Alternatively, manually changing that part of the ansible playbook script to Always pull the image may also work.

**Comment by jekader** [comment]
Thanks, I tried changing the playbook as you recommended but unfortunately that didn't help and etcd to CustomResources still fails with the DC yaml showing the 3.9 image

**Comment by jmontleon** [comment]
If possible can you please share the inventory files for install and upgrade?
